MCC Changes for CLA Backend

imtiazAhmd opened the pull request:
## What does this pull request do?

Introduced a new endpoint `GET /cla_provider/api/v1/case/{case_id}/detailed/` to return case details together with personal information, third-party details, and adaptation data.

This pull request introduces a new `DetailedCaseSerializer` and a corresponding detailed endpoint to the case API, enabling clients to retrieve all nested case details in a single, optimised API call. The changes are designed to be non-breaking, maintainable, and performance-conscious, with comprehensive tests and documentation provided.

### Serializer and API Enhancements

* Added `DetailedCaseSerializer` to `cla_provider/serializers.py`, which extends `CaseSerializer` and includes nested serializers for `personal_details`, `adaptation_details`, `thirdparty_details`, and `eligibility_check`, all set as read-only fields.
* Introduced a new GET endpoint `/cla_provider/api/v1/case/{id}/detailed/` in `CaseViewSet` (`cla_provider/views.py`) using `@detail_route()`, returning a case with all nested details populated.

### Query and Import Optimizations

* Updated imports in `cla_provider/views.py` to include `DetailedCaseSerializer`.
* Enhanced queryset optimization in `CaseViewSet` by including `thirdparty_details` in `select_related` for improved performance.

### Documentation and Testing

* Added a detailed implementation guide in `DETAILED_CASE_SERIALIZER_IMPLEMENTATION.md`, outlining the new serializer, endpoint, API usage, benefits, and performance considerations.
* Provided a comprehensive test script (`test_comprehensive_detailed_serializer.py`) to verify serializer structure, serialization with sample data, and the new API endpoint, along with instructions for manual testing via Postman.

## Any other changes that would benefit highlighting?

After these changes, you'll be able to sort the /case endpoint using both provider_assigned_at and `provider_closed` fields via the `ordering` query parameter, like:

`/case/?ordering=provider_assigned_at`
`/case/?ordering=-provider_closed`


## Checklist

- [ ] Provided JIRA ticket number in the title, e.g. "LGA-152: Sample title"

imtiazAhmd committed:
Create new endpoint for case

imtiazAhmd committed:
Updates auth endpoint to add logged in user details

imtiazAhmd committed:
Adds two fields for ordering case endpoint

imtiazAhmd committed:
Adding coverage for auth view changes to add user info

said-moj committed:
Merge branch 'master' into feature/detailed-case-serializer

imtiazAhmd committed:
Adds case state in the detailed case endpoint

imtiazAhmd committed:
Adds endpoint for marking case as opened

imtiazAhmd committed:
Updates to pass the test

imtiazAhmd committed:
Adds rejected filter for cases endpoint

imtiazAhmd committed:
Adds completed filter for cases endpoint

imtiazAhmd committed:
Adds note associated with states

imtiazAhmd committed:
Adds safe_to_contact field to the case endpoint for search page

imtiazAhmd committed:
Merge remote-tracking branch 'origin/master' into feature/detailed-case-serializer

imtiazAhmd committed:
Addressed Semgrep Appsec warning

imtiazAhmd commented on cla_backend/apps/cla_auth/views.py:
```diff
@@ -150,7 +161,7 @@ def check_login_attempts(self, request):
         cool_off_time = timezone.now() - datetime.timedelta(minutes=settings.LOGIN_FAILURE_COOLOFF_TIME)
 
-        attempts = AccessAttempt.objects.filter(username=username, created__gt=cool_off_time).count()
+        attempts = AccessAttempt.objects.filter(Q(username=username) & Q(created__gt=cool_off_time)).count()
```
/other <The code is written as defensively as possible, We're using explicit Q objects instead of kwargs.
The vulnerability exists in Django itself, not in how we're calling the ORM. For Actual Fix needs to upgrade Django>

imtiazAhmd committed:
Updates test after security patch

imtiazAhmd committed:
Adds phone number to case endpoint for MCC search

imtiazAhmd committed:
Modifies accept endpoint to delete provider_closed for MCC

said-moj committed:
Merge branch 'master' into feature/detailed-case-serializer

willc-work committed:
Merge branch 'master' into feature/detailed-case-serializer

imtiazAhmd committed:
Merge branch 'master' into feature/detailed-case-serializer

imtiazAhmd committed:
Reverts user info from access_token  endpoint

imtiazAhmd committed:
Merge branch 'main' into feature/detailed-case-serializer

imtiazAhmd committed:
Removes readme for detailed endpoint

imtiazAhmd committed:
Adds scope, notes and diagnosis within the detailed endpoint response

imtiazAhmd committed:
Updates case status logic for rejected and completed

imtiazAhmd committed:
Updated case status logic for closed

imtiazAhmd committed:
Reverts back the change for case status closed

imtiazAhmd committed:
Adds a new filter for the list endpoint to accomodate MCC behaviour

imtiazAhmd committed:
Reverts the addition of filter param for list endpoint

imtiazAhmd committed:
Adding the log object rather than just the notes in state_note for detailed endpoint

imtiazAhmd committed:
Adds 'is-urgent' to the detailed endpoint response

imtiazAhmd committed:
Merge branch 'main' into feature/detailed-case-serializer

imtiazAhmd committed:
Adds key to pass the test

imtiazAhmd committed:
Adds providers endpoint to get firm name and allocated categories

imtiazAhmd committed:
Adds state_note to the detailed endpoint to include High level logs

imtiazAhmd committed:
Merge branch 'main' into feature/detailed-case-serializer

imtiazAhmd committed:
Updates test for log

said-moj committed:
Use a perm database for mcc integeration branch (#1364)

said-moj commented:
This commit f06149369ccd64721a8f42d1903bbe86efa15cc3 needs to reverted before merging with main

said-moj commented on cla_backend/apps/cla_provider/serializers.py:
```diff
@@ -296,13 +351,28 @@ class Meta(AdaptationDetailsSerializerBase.Meta):
 
 class StaffSerializer(ExtendedUserSerializerBase):
-    provider = ProviderSerializer(read_only=True)
+    provider = ProviderSerializerMinimal(read_only=True)
```
The `ProviderSerializerMinimal` is missing the `law_category` - was this not being used before?

aaj-moj replied:
Doesn't look like it. Not breaking changes have been seen in MCC.

said-moj replied:
@aaj-moj `law_category` is being used here on the frontend https://github.com/ministryofjustice/cla_frontend/blob/main/cla_frontend/assets-src/javascripts/vendor/calendar.config.js#L81

So if this merged in before providers start using MCC and  they are still using this CHS then this missing field could break the existing frontend.

said-moj committed:
Merge branch 'main' into feature/detailed-case-serializer

MazOneTwoOne committed:
EL-2491: MCC split endpoint (#1357)

If applied, this creates a new splint endpoint for MCC
- moves the SerializerConfigurationTest to test folders
- simplifies how the SplitCaseForm &  SplitMCCCaseForm sub class, work

- EL-3011: MCC: Case flags backend setup for your cases (#1390)

* If applied, this adds the following to the `GET /cla_provider/api/v1/case/` endpoint;
- `vulnerable_user`
- `language`
- `bsl_webcam`
- `text_relay`
- `thirdparty` 
- It also checks if the thirdparty_obj object exists and groups the `mcc_case_flags` together
-  adds a test to check `mcc_case_flags` exist on the serializer

---------

Co-authored-by: said-moj <45761276+said-moj@users.noreply.github.com>

aaj-moj committed:
Revert "Use a perm database for mcc integeration branch (#1364)"

This reverts commit f06149369ccd64721a8f42d1903bbe86efa15cc3.

aaj-moj committed:
fix: revert to objects.get() for proper exceptions when not found

aaj-moj commented on cla_backend/apps/legalaid/models.py:
```diff
@@ -824,7 +824,10 @@ def log_denormalized_outcome_fields(self):
         TODO Replace painfully circuitous reload method when refresh_from_db available in Django 1.8
         """
         if self.id:
-            case = Case.objects.get(id=self.id)
```
@said-moj we have addressed this comment.

aaj-moj commented:
> This commit [f061493](https://github.com/ministryofjustice/cla_backend/commit/f06149369ccd64721a8f42d1903bbe86efa15cc3) needs to reverted before merging with main

@said-moj commit reverted.

aaj-moj committed:
Merge branch 'main' into feature/detailed-case-serializer

aaj-moj committed:
fix: clean up cases at beginning of test

aaj-moj committed:
fix: remove blank line for linting

aaj-moj committed:
fix: make StaffSerializer use ProviderSerializer

aaj-moj committed:
Revert "fix: make StaffSerializer use ProviderSerializer"

This reverts commit e6bbbb02b4769518350401ec9c69736ac25339b5.

aaj-moj committed:
Merge remote-tracking branch 'origin/main' into feature/detailed-case-serializer

aaj-moj committed:
fix: blank line to trigger pipeline again

MazOneTwoOne committed:
EL-3011: MCC: Case flags backend setup for your cases PART 2 (#1395)

* If applied, this updates the conditions for `thirdparty_details` to be true

---------

Co-authored-by: aaj-moj <alberto.aguilerajuarez@justice.gov.uk>

aaj-moj commented:
Closing as this has been merged via https://github.com/ministryofjustice/cla_backend/pull/1398
